proposals/ref-struct-interfaces.md

```diff
@@ -30,7 +31,7 @@ File f = ...;
 IDisposable d = f;
 ```
 
-The ability to implement interfaces is only useful when combined with the ability for `ref struct` to participate in generic arguments (as [laid out later](#ref-struct-generic)).
+The ability to implement interfaces is only useful when combined with the ability for `ref struct` to participate in generic arguments (as [laid out later][ref-struct-generics]).
 
 To allow for interfaces to cover the full expressiveness of a `ref struct`, the language will allow `[UnscopedRef]` to appear on interface methods and properties. When a `ref struct` member implements an interface member with a `[UnscopedRef]` attribute, that `ref struct` member must also be decorated with `[UnscopedRef]`. The attribute is ignored when a `class` or non-ref `struct` implements the interface.
```

Comment:
I'm confused what is happening with `[UnscopedRef]` here, why would non-ref `struct`s ignore it? Does it have the same meaning as normal? Is there any reason why a ref struct implementing member would *have* to have `[UnscopedRef]`?

For example, if I write:
```csharp
interface I1
{
    int FieldAccessor { get; set; }
    [UnscopedRef] ref int RefFieldAccessor { get; } //I assume this is supposed to mean the current meaning - which in this case is that the result may be a memory location on the struct passed by `this` (if a struct obviously)
}

class C1 : I1
{
    int field;
    int FieldAccessor { get => field; set => field = value; }
    ref int RefFieldAccessor => ref field; //ignore [UnscopedRef] as expected (lifetime is always long enough)
}

ref struct R1 : I1
{
    int field;
    int FieldAccessor { get => field; set => field = value; }
    [UnscopedRef] ref int RefFieldAccessor => ref field; //require [UnscopedRef] as expected since it returns a ref to a field, allowed since definition has [UnscopedRef]
}

ref struct R2 : I1
{
    C1 instance;
    int FieldAccessor { get => instance.FieldAccessor; set => instance.FieldAccessor = value; }
    /*[UnscopedRef] - should be optional*/ ref int RefFieldAccessor => ref instance.RefFieldAccessor; //I see no reason for this to require [UnscopedRef] on implementation, since it just means the lifetime is longer than was required by [UnscopedRef]
}

ref struct R2 : I1
{
    ref int fieldRef;
    int FieldAccessor { get => fieldRef; set => fieldRef = value; }
    /*[UnscopedRef]*/ ref int RefFieldAccessor => ref fieldRef; //I see no reason for this to require [UnscopedRef] on implementation, since it just means the lifetime is longer than was required by [UnscopedRef]
}

struct S1 : I1
{
    int field;
    int FieldAccessor { get => field; set => field = value; }
    //why would I not also be able to define this?
    [UnscopedRef] ref int RefFieldAccessor => ref field; //require [UnscopedRef] as expected since it returns a ref to a field, allowed since definition has [UnscopedRef]
}
```

i.e., what I'm suggesting is that to me it would make sense for the rule to be this: interface members can be marked as `[UnscopedRef]`, and interface member implementations in `struct`s or `ref struct`s can only be marked as `[UnscopedRef]` if their definition was also marked `[UnscopedRef]` (and never in `class`es obviously). (simplified to only deal with `this` parameter)

Reply:
The `[UnscopdRef]` annotation is necessary for consumption. Consider as an example: 

```csharp
interface I1 {
    [UnscopedRef] ref P1 { get; } 
    ref P2 { get; } 
}

T M<T>(T p) where T : struct, allows ref struct, I1 {
    // Illegal
    return p.P1;

    // Legal
    return p.P2;
}
```

Just as `[UnscopedRef]` is necessary to distinguish between these two cases when interacting directly with a `ref struct`, it's needed when through an interface that abstracts over a `ref struct`